File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    ArrayList<Integer> al;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            adj.add(new ArrayList<>());
        }

        for (int i = 0; i < prerequisites.length; i++) {
            adj.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }

        int[] visited = new int[numCourses];
        int[] arr = new int[0];
        al = new ArrayList<Integer>();
        for (int i = 0; i < numCourses; i++) {
            if (visited[i] == 0) {
                if (isCyclic(adj, visited, i)) {
                    return arr;
                }
            }
        }
        int[] arrr = new int[al.size()];
        for (int i = 0;i<al.size();i++)
        {
            arrr[i] = al.get(i);
        }
        return arrr;
    }
    
    private boolean isCyclic(List<List<Integer>> adj, int[] visited, int curr) {
        if (visited[curr] == 2) {
            return true;
        }

        visited[curr] = 2;
        for (int i = 0; i < adj.get(curr).size(); i++) {
            if (visited[adj.get(curr).get(i)] != 1) {
                if (isCyclic(adj, visited, adj.get(curr).get(i))) {
                    return true;
                }
            }
        }
        visited[curr] = 1;
        al.add(curr);
        return false;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        prereq = {c: [] for c in range(numCourses)}
        for crs, pre in prerequisites:
            prereq[crs].append(pre)

        output = []
        visit, cycle = set(), set()

        def dfs(crs):
            if crs in cycle:
                return False
            if crs in visit:
                return True

            cycle.add(crs)
            for pre in prereq[crs]:
                if dfs(pre) == False:
                    return False
            cycle.remove(crs)
            visit.add(crs)
            output.append(crs)
            return True

        for c in range(numCourses):
            if dfs(c) == False:
                return []
        return output
        # hashmap = {}
        # visited = set()
        # res = []

        # for depends in prerequisites:
        #     if depends[0] in hashmap:
        #         hashmap[depends[0]].append(depends[1])
        #     else:
        #         hashmap[depends[0]] = [depends[1]]
    
        
        # def ifPossible(i):
        #     if i not in hashmap:
        #         if i not in res:
        #             res.append(i)
        #         return True
        #     if i in visited:
        #         return False
        #     visited.add(i)
        #     for idx in hashmap[i]:
        #         if ifPossible(idx) == False:
        #             return False
        #     visited.remove(i)
        #     if i not in res:
        #         res.append(i)
        #     return True

        # for i in range(numCourses):
        #     if ifPossible(i) == False:
        #         return []
                
        # return res